ILM: parse origination date from index name

andreidan opened the pull request:
Introduce the `index.lifecycle.parse_origination_date` setting that
indicates if the origination date should be parsed from the index name.
If set to true an index which doesn't match the expected format (namely
`indexName-{dateFormat}-optional_digits` will fail before being created.
The origination date will be parsed when initialising a lifecycle for an
index and it will be set as the `index.lifecycle.origination_date` for
that index.

A user set value for `index.lifecycle.origination_date` will always
override a possible parsable date from the index name.

Together with #46561 closes #42449

<!--
Thank you for your interest in and contributing to Elasticsearch! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

andreidan committed:
ILM: parse origination date from index name

elasticmachine commented:
Pinging @elastic/es-core-features

andreidan committed:
Drop unused imports.

andreidan committed:
Use default locale when creating dateformat

dakrone commented:
Thanks for opening this for discussion Andrei! I have some ideas for how to implement it.

> the user simply enables/disables the parsing, without having the option to configure the date format in the index name. Should we be more configurable and allow the user to set the date format? Or even more to provide the full index name pattern? (personal opinion here is to be more restrictive as this feature is not about legacy indexes but newly created ones which must obey the pattern)

I think to start we should support the `yyyy.MM.dd` format that you mentioned. We can always expand later to customization of the format if we have to, but it would be harder to go the other way around.

One thing I notice from your PR is parsing `^.*-(.+)$`, which would support `index-2019.09.18`, but I think we will also want to support `index-2019.09.18-000231` for indices that are using rollover (so the same, just `^.*-(.+)(-[\d]+)?$` I think (I did not double check the regex))

> the way the "parse origination date" and explicit "origination date" interact is that the parse setting has priority if both are configured. I think if we're going this way it should actually be the other way around and have the explicit setting supersede the parse setting. Or maybe having both settings configured should yield an error?

I agree that the `origination_date` set in the settings should always have priority. If it's not set, and `parse_origination_date` *is* set, we should attempt to automatically set the origination date setting to what the parsed value is. We can do this I think in `InitializePolicyContextStep#performAction` when we update the cluster state with the new index metadata (we can update the settings at the same time)

> How lenient should we be to parsing errors (ie. can't find "-" in the index name or the date parsing fails)?

We should be extremely strict about parsing errors, so that we avoid any unintended behavior around the age of the index, especially since this can lead to data being deleted with an unintended retention period.

We can do this by registering an `IndexEventListener` that implements the `beforeIndexCreated` method. In this method, we could validate that the index name correctly matched the regex for parsing the date out of the name. If it doesn't match, we can throw an exception and prevent the index from being created (rather than leniently creating it without parsing the date).

andreidan committed:
ILM: parse origination date from index name

Introduce the `index.lifecycle.parse_origination_date` setting that
indicates if the origination date should be parsed from the index name.
If set to true an index which doesn't match the expected format (namely
`indexName-{dateFormat}-optional_digits` will fail before being created.
The origination date will be parsed when initialising a lifecyle for an
index and it will be set as the `index.lifecycle.origination_date` for
that index.

A user set value for `index.lifecycle.origination_date` will always
override a possible parsable date from the index name.

andreidan committed:
Add license to new file.

andreidan committed:
Rename test class to obbey convention

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-parse-origination-date

andreidan committed:
Register IndexEventListener only if ILM enabled.

andreidan committed:
Use the renamed ilmEnabled flag instead of enabled

andreidan commented:
Adding the initial draft PR description here for completeness (and to provide context to @dakrone 's comment)

Initial draft for parsing the origination date from the index name.

This is meant solely as support for discussion on how to implement
this feature.

A few assumptions that we should discuss:

    the user simply enables/disables the parsing, without having the
    option to configure the date format in the index name. Should we
    be more configurable and allow the user to set the date format? Or
    even more to provide the full index name pattern? (personal opinion
    here is to be more restrictive as this feature is not about legacy
    indexes but newly created ones which must obey the pattern)
    the way the "parse origination date" and explicit "origination date"
    interact is that the parse setting has priority if both are configured.
    I think if we're going this way it should actually be the other way
    around and have the explicit setting supersede the parse setting.
    Or maybe having both settings configured should yield an error?
    How lenient should we be to parsing errors (ie. can't find "-" in the
    index name or the date parsing fails)?

andreidan committed:
Use end of line matcher ($) for index format regex

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-parse-origination-date

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java:
```diff
@@ -34,8 +37,28 @@ public ClusterState performAction(Index index, ClusterState clusterState) {
         LifecycleExecutionState lifecycleState = LifecycleExecutionState
             .fromIndexMetadata(indexMetaData);
+
+        if (shouldParseIndexName(indexMetaData.getSettings())) {
```
I think this can be refactored a bit not to have multiple return paths for the cluster state, instead using a single `ClusterState.Builder` and only adding updated settings iff `shouldParseIndexName` is true

dakrone commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/IndexLifecycleOriginationDateParserTests.java:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        {
+            long parsedDate = parseIndexNameAndExtractDate("indexName-2019.09.04-0000001");
```
Can you add a test for `index-2019.09.04-2018.01.01-002` (or something like it) to make sure it uses the correct date in the name?

andreidan replied:
Adding a test for this but this will currently fail by picking `2019.09.04-2018.01.01` as the date which can't be parsed. Do you think we should do a more extensive validation here? (eg. max 2 dashes in the name)

Adding a test for `indexName-2019.09.04-2019.09.24` to make sure we fetch the correct date (`2019.09.04` in this case)

dakrone replied:
> Adding a test for this but this will currently fail by picking `2019.09.04-2018.01.01` as the date which can't be parsed. Do you think we should do a more extensive validation here? (eg. max 2 dashes in the name)

This seems like a bug then, a user should be free to have any valid index name prior to `-yyyy.MM.dd(-[\\d]+)?`, with as many dashes as they'd like. It seems like for an index named `index-2019.09.04-2018.01.01-002` it should extract `2018.01.01` as the origination date, right?

andreidan replied:
I think you're right, pushing a fix shortly

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleService.java:
```diff
@@ -148,6 +153,13 @@ public String executorName() {
     }
 
+    @Override
+    public void beforeIndexCreated(Index index, Settings indexSettings) {
```
I believe this should be `beforeIndexAddedToCluster` instead, since `beforeIndexCreated` is called on each data node in addition to the master (it isn't a huge deal, but might as well use the one better suited), `beforeIndexAddedToCluster` is only called on the master node.

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycle.java:
```diff
@@ -268,6 +270,13 @@ protected Clock getClock() {
+    @Override
+    public void onIndexModule(IndexModule indexModule) {
+        if (ilmEnabled) {
+            indexModule.addIndexEventListener(indexLifecycleInitialisationService.get());
```
Can you add an `assert indexLifecycleInitialisationService.get() != null` before this line? I want to make sure we don't add any NPEs in the future

dakrone commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/action/TransportExplainLifecycleAction.java:
```diff
@@ -109,7 +109,7 @@ protected void doMasterOperation(ExplainLifecycleRequest request, String[] concr
                 if (request.onlyErrors() == false
                     || (ErrorStep.NAME.equals(lifecycleState.getStep()) || indexLifecycleService.policyExists(policyName) == false)) {
-                    Long originationDate = idxSettings.getAsLong(LIFECYCLE_ORIGINATION_DATE, -1L);
+                    Long originationDate = idxMetadata.getSettings().getAsLong(LIFECYCLE_ORIGINATION_DATE, -1L);
```
Is this change to fix something or was it inadvertent after undoing the other changes here?

andreidan replied:
Yes "manual revert" from the draft. Good catch

dakrone commented on x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleInitialisationTests.java:
```diff
@@ -249,19 +253,86 @@ public void testExplainExecution() throws Exception {
+    public void testExplainParseOriginationDate() throws Exception {
+        // start node
+        logger.info("Starting server1");
+        internalCluster().startNode();
```
Can you start another node (at least 2) to make sure we don't have any issues with master nodes versus data nodes?

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/IndexLifecycleOriginationDateParser.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * Determines if the origination date needs to be parsed from the index name.
+     */
+    public static boolean shouldParseIndexName(Settings indexSettings) {
```
I think this should throw an exception when `origination_date` is set to a value and `parse_origination_date` is set to true. We should be strict and reject the index so the user isn't confused about which one takes priority. What do you think?

andreidan replied:
I initially started going down that road (this method was something along the lines of `validateOriginationDateSettings`) but I chose to avoid it as we will set the `origination_date` setting for the index to the date we parse when we initiate the lifecycle policy. If we are to validate the settings this way, our code is creating an invalid setting combination (which we could potentially address by switching the parse setting to `false` when we set the `origination_date` but I didn't like this as we're changing something the user has explicitly set).

I don't have a strong opinion here, but to me, the semantic of an explicit setting overriding what could be parsed was clearer than changing what the user configured.

dakrone replied:
You make a good point about us internally creating settings that have an invalid combination. Let's leave this as-is then.

dakrone reviewed (changes requested):
Thank Andrei, I think this will be a great feature to have! I left some comments and questions for you on this.

andreidan committed:
Include date in the exception when parsing fails

andreidan committed:
Add more tests around parsing the index name

andreidan committed:
The initialize policy step uses one return path

andreidan committed:
Use beforeIndexAddedToCluster

andreidan committed:
Assert the IndexLifecycleService is not null

andreidan committed:
Use idxSettings directly instead of thru metadata

andreidan committed:
Start 2 nodes in test

andreidan committed:
Allow multiple dates in the index name

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-parse-origination-date

dakrone reviewed (approved):
LGTM, thanks for iterating on this Andrei! I played with it locally and it works very well.
